parsing/parser.mly

```diff
@@ -754,6 +754,32 @@ The precedences must be listed from low to high.
 %inline mkclass(symb): symb
     { mkclass ~loc:$sloc $1 }
 
+/* Generic definitions */
+
+(* [reversed_separated_nonempty_list(separator, X)] recognizes a nonempty list
+   of [X]s, separated with [separator]s, and produces an OCaml list in reverse
+   order -- that is, the last element in the input text appears first in this
+   OCaml list. *)
+
+(* [inline_reversed_separated_nonempty_list(separator, X)] is semantically
+   equivalent to [reversed_separated_nonempty_list(separator, X)], but is
+   marked %inline, which means that the case of a list of length one and
+   the case of a list of length more than one will be distinguished at the
+   use site, and will give rise there to two productions. This can be used
+   to avoid certain conflicts. *)
+
+%inline inline_reversed_separated_nonempty_list(separator, X):
+  x = X
+    { [ x ] }
+| xs = reversed_separated_nonempty_list(separator, X)
```

Comment:
I'm not that familiar with menhir, so I'm going to ask a silly question: is the choice of `reversed_...` over `inline_reversed_...` here intentional (and necessary to avoid a blowup of some sort)?

Reply:
I think that @fpottier used it because the previous `core_type_comma_list` definition returned elements in the reversed order (with `List.rev` in each semantic action). @fpottier, I agree with @trefis' implicit suggestion that using lists that return elements in the right order and removing the `List.rev` at the call site would result in the better grammar, even if it means updating a couple more call-sites.

Reply:
You misunderstood my question: I understand where the `reversed` come from, I was just questioning why in the definition of `inline_reversed_...` we "called" `reversed_...` instead of recursively calling oneself.

Reply:
Indeed, you cannot inline a symbol into its own definition, so the definitions of `reversed_separated_nonempty_list` and `inline_reversed_separated_nonempty_list` have to be written in this style. There is an entry in the Menhir FAQ about this.